DEVELOPER.md

```diff
@@ -0,0 +1,189 @@
+allows you to make use of `pdb` to introspect variables and step through and over code.
+
+Assuming relevant `breakpoint`s have been added, you might debug while testing:
+
+```shell
+$ ./pants test --debug envoy.dependency.check::
+
+```
+
+### Linting the code
+
+The repo is linted with `flake8` and the default config.
+
+To lint the entire repo:
+
+```shell
+$ ./pants lint ::
+
+```
+
+
+## Testing code in an Envoy environment with `bazel`
+
+Ordinarily Envoy uses the pytooling code from its published pypi packages.
+
+When working on code in pytooling it is often helpful to test and introspect changes from
+an Envoy environment without publishing to pypi.
+
+### Setting up the Envoy <> pytooling environment
+
+There are a couple of changes that need to be made in the Envoy environment.
+
+#### Step 1: add dev requirements
+
+You need to set up the dev requirements for Bazel.
+
+For example to work on the `envoy.dependency.check` package, edit `tools/dev/requirements.txt`
+adding the following (adjusted to your environment):
+
+```console
+-e file:///src/workspace/pytooling/envoy.dependency.check#egg=envoy.dependency.check&cachebust=000
+
+```
+
+In this example, my pytooling directory is located in `/src/workspace/pytooling`.
+
+Whatever file path that you put here must be accessible to your Envoy environment, so if you are running inside a container,
+you may need to mount your pytooling directory to a matching path.
+
+Also note the `cachebust` parameter.
+
+When making changes to your pytooling code Bazel does not know to rebuild the package and will default
+to using its cached version. You can change the value of `cachebust` to expire Bazel's cache.
+
+You can also test pytooling (or other Python) code directly from a repo, please see example given
+in `tools/dev/requirements.txt`
+
+#### Step 2: use of dev requirements in Bazel
+
+Much of the pytooling code is accessed by Envoy by way of a Python `entry_point`.
+
+It does this using a macro which wraps the `entry_point` from `rules_python`.
+
+Using the example of dependency check, we can edit the `envoy_entry_point` in `tools/dependency/BUILD`.
+
+Firstly, at the top of the file, add:
+
+```starlark
+
+load("@dev_pip3//:requirements.bzl", dev_entry_point = "entry_point")
+
+```
+
+Now we want to update the `check` rule to add the `dev_entry_point`:
+
+```starlark
+
+envoy_entry_point(
+    name = "check",
+    entry_point = dev_entry_point,
+    args = [
+        "--repository_locations=$(location //bazel:all_repository_locations)",
+        "--cve_config=$(location :cve.yaml)",
+    ],
+    data = [
+        ":cve.yaml",
+        "//bazel:all_repository_locations",
+    ],
+    pkg = "envoy.dependency.check",
+)
+
+```
+
+### Running the development code
+
+With the development environment setup we can now run the code from Envoy Bazel.
+
+Assuming we want to test the dependency check rule, as configured in the above example:
+
+```console
+
+$ bazel run //tools/dependency:check
```

Comment:
may be we can add an example on how to pass arguments to show how to run a single check?